### PR TITLE
Estimate deploy account fee

### LINKED ISSUE #298: Allow fee estimation for deploy account tx

### PR DESCRIPTION
**Pull request by Nathan-SL**
## Usage related changes

-   The `estimateDeployAccountFee()` function can be used to obtain the account deployment fee.
-   Closes #298 

## Development related changes

-   NA

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [x] Documented the changes
-   [ ] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [x] Linked issues which this PR resolves
-   [x] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example):https://github.com/Shard-Labs/starknet-hardhat-example/pull/96
    -   [x] Modified `test.sh` to use my example repo branch
    -   [x] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [ ] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

File: src/account.ts [hunk 1]
```diff
@@ -757,6 +795,50 @@ export class ArgentAccount extends Account {
...
+
+        const signature = this.getSignatures(msgHash);
+        const data = {
+            type: "DEPLOY_ACCOUNT",
+            class_hash: ArgentAccount.PROXY_CLASS_HASH,
+            constructor_calldata: constructorCalldata,
+            contract_address_salt: this.salt,
+            signature: bnToDecimalStringArray(signature || []),
+            version: numericToHexString(QUERY_VERSION),
+            nonce: numericToHexString(0)
```
Comment (FabijanC) [inline src/account.ts:836]: Why not use the defined `nonce` variable?

File: src/account.ts [hunk 2]
```diff
@@ -548,11 +535,62 @@ export class OpenZeppelinAccount extends Account {
...
+            this.salt,
+            ...constructorCalldata
+        ]);
+        const msgHash = hash.computeHashOnElements([
+            TransactionHashPrefix.DEPLOY_ACCOUNT,
+            numericToHexString(QUERY_VERSION),
+            this.address,
+            0, // entrypoint selector is implied
+            calldataHash,
+            maxFee,
```
Comment (FabijanC) [inline src/account.ts:560]: Does estimation really require a maxFee? Or is some other value expected to be zero here?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/account.ts (hunk 1 above)
Comment (Nathan-SL) [reply src/account.ts]: https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/304/commits/eac6546509a07710b6ef2776f4f66b6e4f8db708

File: src/account.ts (hunk 2 above)
Comment (Nathan-SL) [reply src/account.ts:560]: Yes, I think I linked to the wrong function in the prev comment, but if you see the `estimateAccountDeployFee()` [here](https://github.com/0xs34n/starknet.js/blob/47ffeb6a611eb425482741aaca27aa444a9cd0ea/src/account/default.ts#L196) the estimation also passes a Zero `maxFee` to get a signature. And to clear things out, the `estimateDeployAccountFee()` is looking at the estimation.

**Pull request merged**